src/System Application/App/AppSource Gallery/src/AppSourceProductList.Page.al

```diff
@@ -0,0 +1,208 @@
+        }
+
+        area(Processing)
+        {
+            action(OpenAppSource)
+            {
+                Caption = 'View AppSource';
+                Scope = Repeater;
+                Image = OpenWorksheet;
+                ToolTip = 'View all apps in AppSource';
+
+                trigger OnAction()
+                begin
+                    AppSourceProductManager.OpenAppSource();
+                end;
+            }
+
+            action(OpenInAppSource)
+            {
+                Caption = 'View in AppSource';
+                Scope = Repeater;
+                Image = Open;
+                ToolTip = 'View selected app in AppSource';
+
+                trigger OnAction()
+                var
+                    SelectedRec: Record "AppSource Product";
+                begin
+                    CurrPage.SetSelectionFilter(SelectedRec);
+                    SelectedRec.Next();
+
+                    if (SelectedRec.Count() = 1) then
+                        AppSourceProductManager.OpenInAppSource(SelectedRec.UniqueProductID)
+                    else
+                        Error(SelectOneRowErrLbl);
+                end;
+            }
+
+            action(ShowSettings)
+            {
+                Caption = 'Edit User Settings';
+                RunObject = Page "User Settings";
+                Image = UserSetup;
+                ToolTip = 'Locale will be used to determine the market and language will be used to determine the language of the app details listed here.';
+            }
+        }
+
+        area(Navigation)
+        {
+            action(UpdateProducts)
+            {
+                Caption = 'Refresh list from Microsoft AppSource';
+                Scope = Page;
+                ToolTip = 'Refreshes the list by downloading the latest apps from Microsoft AppSource';
+                Image = Refresh;
+
+                trigger OnAction()
+                begin
+                    Rec.ReloadAllProducts();
+                    CurrPage.Update();
+                end;
+
+            }
+        }
+    }
+
+    views
+    {
+        view("TopRated")
+        {
+            Caption = 'Top rated apps';
+            Filters = where(RatingAverage = filter(>= 4));
+            OrderBy = descending(RatingAverage);
+        }
+
+        view("Popular")
+        {
+            Caption = 'Popular apps';
+            OrderBy = descending(Popularity);
+        }
+
+        view("Resent Updates")
+        {
+            Caption = 'Recently changed apps';
+            OrderBy = descending(LastModifiedDateTime);
+        }
+    }
+
+    var
+        ExtensionManagement: Codeunit "Extension Management";
+        AppSourceProductManager: Codeunit "AppSource Product Manager";
+        CurrentRecordCanBeUninstalled: Boolean;
+        SelectOneRowErrLbl: Label 'Action requires exactly one row to be selected.';
+
+    trigger OnOpenPage()
+    begin
+        rec.ReloadAllProducts();
```

Comment:
Team, this will freeze the UI for tens of seconds or even minutes for the amount of apps nearly 5000. And assume the volume of API calls if several users will do that in parallel. I would advise to refactor this, at least update once when installing the app, with the job queue to update periodically and manual action to update on demand. This will not work for the users onOpenPage

Reply:
<img width="1017" alt="image" src="https://github.com/microsoft/BCApps/assets/29101017/c3740382-a0ef-4088-81e3-158027d4704d">
So average speed is 2.76 sec for 1000 records, would result in about 13-15 sec just to get the list of all apps + time to parse json and populate to the temp table, approximately 20-25 sec of user waiting every time user opens the page

Reply:
Next PR